File: atlas-ai-py/atlas_ai/optimization/deep_optimizer.py
import math
import random
from copy import deepcopy
from typing import Callable


class OptimizationResult:
    def __init__(self):
        self.optimal_solution: list[float] = []
        self.optimal_value: float = 0
        self.iterations: int = 0
        self.converged: bool = False
        self.history: list[float] = []
        self.message: str = ""


class OptimizationConfig:
    def __init__(self):
        self.objective: Callable = lambda s: sum(v * v for v in s)
        self.constraints: list[Callable] = []
        self.bounds: list[tuple[float, float]] = []
        self.max_iterations: int = 1000
        self.tolerance: float = 1e-6
        self.learning_rate: float = 0.01


class OptimizationState:
    def __init__(self):
        self.current_solution: list[float] = []
        self.current_value: float = 0
        self.iteration: int = 0
        self.gradient: list[float] = []
        self.best_solution: list[float] = []
        self.best_value: float = 0


class DeepOptimizer:
    def __init__(self, config: dict | None = None):
        if config is None:
            config = {}
        self.objective: Callable = config.get("objective", lambda s: sum(v * v for v in s))
        self.constraints: list[Callable] = list(config.get("constraints", []))
        self.bounds: list[tuple[float, float]] = list(config.get("bounds", []))
        self.max_iterations: int = config.get("maxIterations", 1000)
        self.tolerance: float = config.get("tolerance", 1e-6)
        self.learning_rate: float = config.get("learningRate", 0.01)

    def optimize(self, initial_solution: list[float]) -> OptimizationResult:
        history: list[float] = []
        state = self._initialize_state(initial_solution)
        converged = False
        message = ""

        for iteration in range(self.max_iterations):
            state.iteration = iteration

            state.gradient = self._compute_gradient(state.current_solution)

            state.current_solution = self._update_solution(state.current_solution, state.gradient)

            state.current_solution = self._apply_constraints(state.current_solution)

            state.current_value = self.objective(state.current_solution)
            history.append(state.current_value)

            if state.current_value < state.best_value:
                state.best_solution = list(state.current_solution)
                state.best_value = state.current_value

            if self._check_convergence(state, iteration):
                converged = True
                message = f"Converged after {iteration + 1} iterations"
                break

        if not converged:
            message = f"Max iterations ({self.max_iterations}) reached"

        result = OptimizationResult()
        result.optimal_solution = state.best_solution
        result.optimal_value = state.best_value
        result.iterations = state.iteration + 1
        result.converged = converged
        result.history = history
        result.message = message
        return result

    def evaluate(self, solution: list[float]) -> float:
        evaluated = list(solution)
        if len(self.bounds) > 0:
            evaluated = [
                max(self.bounds[i][0], min(self.bounds[i][1], val))
                if i < len(self.bounds) else val
                for i, val in enumerate(solution)
            ]
        return self.objective(evaluated)

    def converge(self, current_value: float, previous_value: float) -> bool:
        return abs(previous_value - current_value) < self.tolerance

    def set_objective(self, objective: Callable):
        self.objective = objective

    def add_constraint(self, constraint: Callable):
        self.constraints.append(constraint)

    def set_bounds(self, bounds: list[tuple[float, float]]):
        self.bounds = list(bounds)

    def set_learning_rate(self, rate: float):
        self.learning_rate = max(0.0001, min(1.0, rate))

    def get_config(self) -> dict:
        return {
            "objective": self.objective,
            "constraints": list(self.constraints),
            "bounds": list(self.bounds),
            "maxIterations": self.max_iterations,
            "tolerance": self.tolerance,
            "learningRate": self.learning_rate,
        }

    def reset(self):
        self.constraints = []
        self.bounds = []
        self.max_iterations = 1000
        self.tolerance = 1e-6
        self.learning_rate = 0.01

    def optimize_with_restarts(
        self, initial_solution: list[float], num_restarts: int = 5
    ) -> OptimizationResult:
        best_result: OptimizationResult | None = None

        for i in range(num_restarts):
            perturbed = [v + (random.random() - 0.5) * 0.1 for v in initial_solution]
            result = self.optimize(perturbed)

            if best_result is None or result.optimal_value < best_result.optimal_value:
                best_result = result

            if best_result is not None and best_result.optimal_value < self.tolerance:
                break

        if best_result is None:
            return self.optimize(initial_solution)
        return best_result

    def optimize_coordinate_descent(self, initial_solution: list[float]) -> OptimizationResult:
        history: list[float] = []
        solution = list(initial_solution)
        best_value = self.objective(solution)
        best_solution = list(solution)
        converged = False
        message = ""

        for iteration in range(self.max_iterations):
            improved = False

            for i in range(len(solution)):
                original_value = solution[i]

                step_size = self.learning_rate * (1 + iteration * 0.01)

                for delta in (step_size, -step_size):
                    solution[i] = original_value + delta
                    solution = self._apply_constraints(solution)

                    new_value = self.objective(solution)

                    if new_value < best_value:
                        best_value = new_value
                        best_solution = list(solution)
                        improved = True
                    else:
                        solution[i] = original_value

            history.append(best_value)

            if not improved or best_value < self.tolerance:
                converged = True
                message = f"Converged after {iteration + 1} iterations"
                break

        if not converged:
            message = f"Max iterations ({self.max_iterations}) reached"

        result = OptimizationResult()
        result.optimal_solution = best_solution
        result.optimal_value = best_value
        result.iterations = self.max_iterations
        result.converged = converged
        result.history = history
        result.message = message
        return result

    def _initialize_state(self, initial_solution: list[float]) -> OptimizationState:
        current_value = self.objective(initial_solution)
        state = OptimizationState()
        state.current_solution = list(initial_solution)
        state.current_value = current_value
        state.iteration = 0
        state.gradient = [0.0] * len(initial_solution)
        state.best_solution = list(initial_solution)
        state.best_value = current_value
        return state

    def _compute_gradient(self, solution: list[float]) -> list[float]:
        gradient: list[float] = []
        epsilon = 1e-6

        for i in range(len(solution)):
            original = solution[i]

            def make_plus():
                orig = original
                idx = i
                def f(vals):
                    vals2 = list(vals)
                    vals2[idx] = orig + epsilon
                    return vals2
                return f

            plus_vals = list(solution)
            plus_vals[i] = original + epsilon
            plus = self.objective(plus_vals)

            minus_vals = list(solution)
            minus_vals[i] = original - epsilon
            minus = self.objective(minus_vals)

            gradient.append((plus - minus) / (2 * epsilon))

        return gradient

    def _update_solution(self, solution: list[float], gradient: list[float]) -> list[float]:
        return [v - self.learning_rate * g for v, g in zip(solution, gradient)]

    def _apply_constraints(self, solution: list[float]) -> list[float]:
        constrained = list(solution)

        for i in range(len(constrained)):
            if i < len(self.bounds):
                constrained[i] = max(
                    self.bounds[i][0], min(self.bounds[i][1], constrained[i])
                )

        for constraint in self.constraints:
            if not constraint(constrained):
                constrained = self._adjust_for_constraint(constrained, constraint)

        return constrained

    def _adjust_for_constraint(
        self, solution: list[float], constraint: Callable
    ) -> list[float]:
        adjusted = list(solution)
        max_attempts = 100

        for attempt in range(max_attempts):
            if constraint(adjusted):
                break
            for i in range(len(adjusted)):
                if i < len(self.bounds):
                    bounds_range = self.bounds[i][1] - self.bounds[i][0]
                    adjusted[i] += (random.random() - 0.5) * bounds_range * 0.1
                else:
                    adjusted[i] += (random.random() - 0.5) * 0.1

            for i in range(len(adjusted)):
                if i < len(self.bounds):
                    adjusted[i] = max(
                        self.bounds[i][0], min(self.bounds[i][1], adjusted[i])
                    )

        return adjusted

    def _check_convergence(self, state: OptimizationState, iteration: int) -> bool:
        gradient_magnitude = math.sqrt(sum(g * g for g in state.gradient))
        if gradient_magnitude < self.tolerance:
            return True

        if state.best_value < self.tolerance:
            return True

        return False

File: atlas-ai/Optimization/DeepOptimizer.ts
/**
 * DeepOptimizer - Advanced optimization component
 * Supports various optimization strategies and convergence detection
 */

export interface OptimizationResult {
  optimalSolution: number[];
  optimalValue: number;
  iterations: number;
  converged: boolean;
  history: number[];
  message: string;
}

export interface OptimizationConfig {
  objective: (solution: number[]) => number;
  constraints?: Array<(solution: number[]) => boolean>;
  bounds?: Array<[number, number]>;
  maxIterations: number;
  tolerance: number;
  learningRate?: number;
}

export interface OptimizationState {
  currentSolution: number[];
  currentValue: number;
  iteration: number;
  gradient: number[];
  bestSolution: number[];
  bestValue: number;
}

/**
 * DeepOptimizer - Advanced optimization with gradient descent and constraint handling
 */
export class DeepOptimizer {
  private objective: (solution: number[]) => number;
  private constraints: Array<(solution: number[]) => boolean>;
  private bounds: Array<[number, number]>;
  private maxIterations: number;
  private tolerance: number;
  private learningRate: number;

  constructor(config: Partial<OptimizationConfig> = {}) {
    this.objective = config.objective || ((s) => s.reduce((a, b) => a + b * b, 0));
    this.constraints = config.constraints || [];
    this.bounds = config.bounds || [];
    this.maxIterations = config.maxIterations || 1000;
    this.tolerance = config.tolerance || 1e-6;
    this.learningRate = config.learningRate || 0.01;
  }

  /**
   * Main optimization method - finds optimal solution
   */
  optimize(initialSolution: number[]): OptimizationResult {
    const history: number[] = [];
    let state = this.initializeState(initialSolution);
    let converged = false;
    let message = "";

    for (let iter = 0; iter < this.maxIterations; iter++) {
      state.iteration = iter;

      // Calculate gradient
      state.gradient = this.computeGradient(state.currentSolution);

      // Update solution using gradient descent
      state.currentSolution = this.updateSolution(state.currentSolution, state.gradient);

      // Apply constraints and bounds
      state.currentSolution = this.applyConstraints(state.currentSolution);

      // Evaluate new solution
      state.currentValue = this.objective(state.currentSolution);
      history.push(state.currentValue);

      // Track best solution
      if (state.currentValue < state.bestValue) {
        state.bestSolution = [...state.currentSolution];
        state.bestValue = state.currentValue;
      }

      // Check convergence
      if (this.checkConvergence(state, iter)) {
        converged = true;
        message = `Converged after ${iter + 1} iterations`;
        break;
      }
    }

    if (!converged) {
      message = `Max iterations (${this.maxIterations}) reached`;
    }

    return {
      optimalSolution: state.bestSolution,
      optimalValue: state.bestValue,
      iterations: state.iteration + 1,
      converged,
      history,
      message,
    };
  }

  /**
   * Evaluates the objective function for a given solution
   */
  evaluate(solution: number[]): number {
    // Apply bounds if defined
    let evaluated = solution;
    if (this.bounds.length > 0) {
      evaluated = solution.map((val, i) => {
        if (this.bounds[i]) {
          return Math.max(this.bounds[i][0], Math.min(this.bounds[i][1], val));
        }
        return val;
      });
    }

    return this.objective(evaluated);
  }

  /**
   * Checks if the optimization has converged
   */
  converge(currentValue: number, previousValue: number): boolean {
    if (Math.abs(previousValue - currentValue) < this.tolerance) {
      return true;
    }
    return false;
  }

  /**
   * Sets a new objective function
   */
  setObjective(objective: (solution: number[]) => number): void {
    this.objective = objective;
  }

  /**
   * Adds a constraint
   */
  addConstraint(constraint: (solution: number[]) => boolean): void {
    this.constraints.push(constraint);
  }

  /**
   * Sets bounds for variables
   */
  setBounds(bounds: Array<[number, number]>): void {
    this.bounds = bounds;
  }

  /**
   * Sets the learning rate
   */
  setLearningRate(rate: number): void {
    this.learningRate = Math.max(0.0001, Math.min(1.0, rate));
  }

  /**
   * Gets current configuration
   */
  getConfig(): OptimizationConfig {
    return {
      objective: this.objective,
      constraints: [...this.constraints],
      bounds: [...this.bounds],
      maxIterations: this.maxIterations,
      tolerance: this.tolerance,
      learningRate: this.learningRate,
    };
  }

  /**
   * Resets the optimizer to initial state
   */
  reset(): void {
    this.constraints = [];
    this.bounds = [];
    this.maxIterations = 1000;
    this.tolerance = 1e-6;
    this.learningRate = 0.01;
  }

  /**
   * Performs multiple optimization runs with random restarts
   */
  optimizeWithRestarts(
    initialSolution: number[],
    numRestarts: number = 5
  ): OptimizationResult {
    let bestResult: OptimizationResult | null = null;

    for (let i = 0; i < numRestarts; i++) {
      // Add small random perturbation to initial solution
      const perturbedSolution = initialSolution.map(
        (val) => val + (Math.random() - 0.5) * 0.1
      );

      const result = this.optimize(perturbedSolution);

      if (!bestResult || result.optimalValue < bestResult.optimalValue) {
        bestResult = result;
      }

      // Early termination if near-optimal solution found
      if (bestResult && bestResult.optimalValue < this.tolerance) {
        break;
      }
    }

    return bestResult || this.optimize(initialSolution);
  }

  /**
   * Performs coordinate descent optimization
   */
  optimizeCoordinateDescent(initialSolution: number[]): OptimizationResult {
    const history: number[] = [];
    let solution = [...initialSolution];
    let bestValue = this.objective(solution);
    let bestSolution = [...solution];
    let converged = false;
    let message = "";

    for (let iter = 0; iter < this.maxIterations; iter++) {
      let improved = false;

      // Optimize each coordinate sequentially
      for (let i = 0; i < solution.length; i++) {
        const originalValue = solution[i];

        // Try positive and negative directions
        const stepSize = this.learningRate * (1 + iter * 0.01);

        for (const delta of [stepSize, -stepSize]) {
          solution[i] = originalValue + delta;
          solution = this.applyConstraints(solution);

          const newValue = this.objective(solution);

          if (newValue < bestValue) {
            bestValue = newValue;
            bestSolution = [...solution];
            improved = true;
          } else {
            solution[i] = originalValue; // Revert
          }
        }
      }

      history.push(bestValue);

      // Check convergence
      if (!improved || bestValue < this.tolerance) {
        converged = true;
        message = `Converged after ${iter + 1} iterations`;
        break;
      }
    }

    if (!converged) {
      message = `Max iterations (${this.maxIterations}) reached`;
    }

    return {
      optimalSolution: bestSolution,
      optimalValue: bestValue,
      iterations: this.maxIterations,
      converged,
      history,
      message,
    };
  }

  /**
   * Initializes the optimization state
   */
  private initializeState(initialSolution: number[]): OptimizationState {
    const currentValue = this.objective(initialSolution);

    return {
      currentSolution: [...initialSolution],
      currentValue,
      iteration: 0,
      gradient: new Array(initialSolution.length).fill(0),
      bestSolution: [...initialSolution],
      bestValue: currentValue,
    };
  }

  /**
   * Computes numerical gradient
   */
  private computeGradient(solution: number[]): number[] {
    const gradient: number[] = [];
    const epsilon = 1e-6;

    for (let i = 0; i < solution.length; i++) {
      const original = solution[i];

      const plus = this.objective(
        solution.map((v, j) => (j === i ? original + epsilon : v))
      );
      const minus = this.objective(
        solution.map((v, j) => (j === i ? original - epsilon : v))
      );

      gradient.push((plus - minus) / (2 * epsilon));
    }

    return gradient;
  }

  /**
   * Updates solution using gradient descent
   */
  private updateSolution(solution: number[], gradient: number[]): number[] {
    return solution.map((val, i) => {
      const delta = this.learningRate * gradient[i];
      return val - delta;
    });
  }

  /**
   * Applies constraints to a solution
   */
  private applyConstraints(solution: number[]): number[] {
    let constrained = [...solution];

    // Apply bounds
    for (let i = 0; i < constrained.length; i++) {
      if (this.bounds[i]) {
        constrained[i] = Math.max(
          this.bounds[i][0],
          Math.min(this.bounds[i][1], constrained[i])
        );
      }
    }

    // Apply constraint functions (try to satisfy)
    for (const constraint of this.constraints) {
      if (!constraint(constrained)) {
        // Try to adjust solution to satisfy constraint
        constrained = this.adjustForConstraint(constrained, constraint);
      }
    }

    return constrained;
  }

  /**
   * Adjusts solution to satisfy a constraint
   */
  private adjustForConstraint(
    solution: number[],
    constraint: (s: number[]) => boolean
  ): number[] {
    let adjusted = [...solution];
    const maxAttempts = 100;

    for (let attempt = 0; attempt < maxAttempts && !constraint(adjusted); attempt++) {
      // Add small random perturbations
      adjusted = adjusted.map((val, i) => {
        if (this.bounds[i]) {
          return val + (Math.random() - 0.5) * (this.bounds[i][1] - this.bounds[i][0]) * 0.1;
        }
        return val + (Math.random() - 0.5) * 0.1;
      });

      // Reapply bounds
      for (let i = 0; i < adjusted.length; i++) {
        if (this.bounds[i]) {
          adjusted[i] = Math.max(
            this.bounds[i][0],
            Math.min(this.bounds[i][1], adjusted[i])
          );
        }
      }
    }

    return adjusted;
  }

  /**
   * Checks convergence criteria
   */
  private checkConvergence(state: OptimizationState, iteration: number): boolean {
    // Check if gradient magnitude is small
    const gradientMagnitude = Math.sqrt(
      state.gradient.reduce((sum, g) => sum + g * g, 0)
    );
    if (gradientMagnitude < this.tolerance) {
      return true;
    }

    // Check if value change is small
    if (state.iteration > 0 && iteration > 0) {
      // This would need access to previous value, simplified here
    }

    // Check if solution is near optimal (value is small for minimization)
    if (state.bestValue < this.tolerance) {
      return true;
    }

    return false;
  }
}